CIP-0020/CIP-0020.md

```diff
@@ -0,0 +1,105 @@
+
+## Motivation
+
+It’s desirable to have interoperability between different transaction signing tools (e.g. cardano-cli, HW wallets, cardano-hw-cli), some of which may need to serialise the transaction by itselves (especially HW wallets, due to security/performance considerations). By interoperability we mean that given the same inputs (transaction parameters) the tools would produce the same outputs (e.g. transaction hash).
+
+This is particularly relevant with the introduction of multisig. If the party initially creating the transaction uses e.g. cardano-cli to create the transaction, then other involved multisig parties might not be able to sign the transaction using HW wallets, because HW wallets might order some elements in the transaction differently thus they would come up with a different transaction hash. The same applies when the transaction is initially created by a HW wallet and requires signatures from parties using cardano-cli.
+
+For example, in order to sign a transaction with a HW wallet, cardano-hw-cli takes the serialised transaction body from cardano-cli, parses it and transforms it into input parameters for the HW wallet. Without canonical CBOR serialisation it might happen that the HW wallet will calculate a different transaction hash than cardano-cli and the signature returned would thus be invalid for the transaction created by cardano-cli.
+
+Another motivation for this CIP is that without canonical CBOR serialisation, specifically canonical map key order, HW wallets would not be able to verify that multi-asset policies and asset_names are included only once in a given output or policy respectively, as they don't have the memory to store all previously passed tokens for the sake of deduplication. This might then cause the user to sign a transaction with the same token included twice, although at the protocol level, only one use of the token would be accounted for.
+
+## Specification
+
+This specification is a modification of the canonical CBOR suggestions taken from the [CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9). Parts not mentioned in this CIP should follow the canonical CBOR suggestions from the CBOR RFC.
+
+### Map key ordering
+
+Keys in a map should be ordered lexicographically based on their real, logical value, not their CBOR representation. If the keys are of different types then the order described below should be maintained.
+
+**Ordering of keys with different types**
+
+The order of types is based on the CBOR major type’s 3 bit header identificator:
+
+1. unsigned integer - major type 0
+2. negative integer - major type 1
+3. byte string - major type 2
+4. text string - major type 3
+5. floating-point numbers - major type 7
+
+I.e. first come the uints, then the ints, byte strings, text strings and floating-point numbers.
```

Comment:
I am afraid this would clash with what cardano-cli authors consider "logical". I.e. if we have only integers as keys (including negative), why include the positive ones first?...

Reply:
Hm. That makes sense. I just don't know what the right approach is then. What if floating-point numbers came into the mix? Perhaps numbers should come first and be ordered by size no matter the type e.g.:

```
-123, -20.7, 30, 30.5
```

Would that make sense? And after that would come bytes and then strings. I.e.:

```
-123, -20.7, 30, 30.5, h'1010', h'20', 'aa', 'b'
```

@refi93 what do you think?

Reply:
I think this decision should be left to IOHK. I don't think floating point numbers will ever be used (notice how pool margin is dealt with).

But whatever is written here, it must correspond to tx body map key ordering.